src/loaders/compute_tools/microtrait/microtrait_util.py

```diff
@@ -20,9 +23,53 @@
     _UNWRAPPED_RULE_COL]
 
 
+def _retrieve_trait_substrate_mapping(
+        substrate2rule_file: Path
+) -> dict[str, set[str]]:
+    # retrieve the trait to substrate mapping from the substrate2rule file
+    substrate2rule_df = pd.read_csv(substrate2rule_file, sep='\t')
+
+    # create a dictionary of trait names to substrate names from the substrate2rule file
+    trait_substrate_mapping = defaultdict(set)
+    for index, row in substrate2rule_df.iterrows():
+        substrate_name = row[_SUBSTRATE2RULE_SUBSTRATE_NAME_COL]
+        for trait_name_col in _RULE2TRAIT_TRAIT_NAME_COLS:
+            trait_name = row[trait_name_col]
+            trait_substrate_mapping[trait_name].add(substrate_name) if pd.notna(trait_name) else None
```

Comment:
> `pd.notna(trait_name)`

Does this happen?

Reply:
not for substrate file. updated to raise if this happens.

Reply:
:+1: